### PR TITLE
(FACT-1663) Add Xen detector

### PR DESCRIPTION
**Pull request by caseywilliams**
The first commit changes the CPUID data source so that it's possible to supply a value for the ecx register (a subleaf) in addition to the eax leaf. This is required to get vendor and version information for Xen PV guests.

The second commit adds the Xen detector, which is based on Xen's [xen-detect.c](https://raw.githubusercontent.com/mirage/xen/master/tools/misc/xen-detect.c). To get a Xen vendor string out of CPUID in a PV context, one has to use a forced emulation prefix ([see here](https://xenbits.xen.org/docs/unstable/features/feature-levelling.html#xen-forced-emulation-prefix)), so there's a new specialized source for Xen (the drop in coverage comes from this addition).

For a valid result, metadata  includes:
- A `context` string ("pv" or "hvm")
- A `version` string, e.g. "4.8"
- A `privileged` bool, which is true for dom0 (this is not based on CPUID - instead it checks the `/proc/xen/capabilities` file, which contains 'control_d' on dom0)

Results from this detector match the results of `xen-detect.c` for PV and HVM Linux and Windows machines I tested manually, but I don't have a lot of Xen experience - advice is very welcome!

### CODE REVIEW

**Review by scottsolowild — APPROVED** [review]

**Comment by branan** [comment]
Wow, that Xen detection is doing some crazy stuff with SIGILL. I want to try to wrap my brain around this a bit more before I hit the merge button O_O

**Comment by caseywilliams** [comment]
I've updated this PR so that it uses the same detection logic as virt-what. It now does the following:

- Checks for a `/proc/xen/` directory and immediately returns if there isn't one (all Xen domains should have this)
- Checks for a cpuid vendor value of `XenVMMXenVMM` - only HVM has this, so this vendor isn't found, it's PV.
- Checks for a `capabilities` file in `/proc/xen/` - if this exists and contains "control_d", it's dom0, otherwise it's domU.

This gets around some of the uglier usage of Xen's PV CPUID instruction, but it removes the version metadata. I'm sure this is something that could be added back in later, when there's more time to carefully consider how it's done.

**Review by branan — APPROVED** [review]
This looks much saner to me now, and the coverage drop seems aceptable :+1:

**Pull request merged**
